From: Mikhael Goikhman <migo@mellanox.com>
Subject: [PATCH] BACKPORT: drivers/infiniband/core/restrack.c

---
 drivers/infiniband/core/restrack.c | 1 +
 1 file changed, 1 insertion(+)

--- a/drivers/infiniband/core/restrack.c
+++ b/drivers/infiniband/core/restrack.c
@@ -339,6 +339,7 @@ void rdma_restrack_del(struct rdma_restr
 	rdma_restrack_put(res);
 	wait_for_completion(&res->comp);
 
+
 out:
 	if (res->task) {
 		put_task_struct(res->task);
